File: examples/ruby/intent-output-p2pkh/IntentOutputP2PKH.runar.rb
require 'runar'

# IntentOutputP2PKH -- exercises the requireOutputP2PKH intent intrinsic.
#
# Note: the property bondPKH and the builtin requireOutputP2PKH are
# spelled in camelCase here (rather than the usual Ruby snake_case)
# because the all-caps PKH token does not round-trip through the
# naive snake_case <-> camelCase transform shared across the
# cross-tier Ruby parsers. Writing them in their canonical camelCase
# form avoids the lossy conversion and keeps the AST byte-identical
# to the Go-source AST after the Go parser lowercases the leading
# char of BondPKH / RequireOutputP2PKH.
class IntentOutputP2PKH < Runar::StatefulSmartContract
  prop :bondPKH, ByteString, readonly: true
  prop :bondAmount, Bigint, readonly: true
  prop :count, Bigint

  def initialize(bondPKH, bondAmount, count)
    super(bondPKH, bondAmount, count)
    @bondPKH = bondPKH
    @bondAmount = bondAmount
    @count = count
  end

  runar_public
  def payBond
    requireOutputP2PKH(0, @bondPKH, @bondAmount)
    @count = @count + 1
  end
end

File: examples/zig/intent-output-p2pkh/IntentOutputP2PKH.runar.zig
const runar = @import("runar");

pub const IntentOutputP2PKH = struct {
    pub const Contract = runar.StatefulSmartContract;

    bondPKH: runar.ByteString,
    bondAmount: i64,
    count: i64 = 0,

    pub fn init(bondPKH: runar.ByteString, bondAmount: i64, count: i64) IntentOutputP2PKH {
        return .{ .bondPKH = bondPKH, .bondAmount = bondAmount, .count = count };
    }

    pub fn payBond(self: *IntentOutputP2PKH) void {
        runar.requireOutputP2PKH(0, self.bondPKH, self.bondAmount);
        self.count = self.count + 1;
    }
};
